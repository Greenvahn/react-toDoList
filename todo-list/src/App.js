
import React, { useState } from 'react';
import Header from './components/Header';
import ToDoList from './components/ToDoList';
import data from './rawData/data';
import './App.css';

function App() {
  const [ toDoList ] = useState(data);
  
  return (
    <div className="App">
      <Header />
      <ToDoList listItems={toDoList} />
    </div>
  );
}

export default App;
