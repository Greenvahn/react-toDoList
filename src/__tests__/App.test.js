import React from "react";
import { fireEvent, render, screen } from "@testing-library/react";
import data from '../rawData/data.json';
import App from "../App"

describe('<App.js />', () => {
  it('the task data is correct', () => {
    expect(data).toMatchSnapshot();
    expect(data).toHaveLength(data.length);
    expect(data.map(item => item.task)).toEqual([
      'Give dog a bath',
      'Do laundry',
      'Vacuum floor',
      'Feed cat',
      'Change light bulbs'
    ]);
  });

  
  it('Renders correctly', () => {

    // Arrange
    render(<App />)

    // Assert
    const items = screen.getAllByRole("listitem")
    const submit_button = screen.getByTestId("submit-btn")
    const list = screen.getByRole("list")
    const form = screen.getByTestId('form')
    const formInput = screen.getByTestId('input')

    // Expect
    expect(list).toBeInTheDocument();
    expect(items).not.toHaveLength(0)
    expect(submit_button).toBeInTheDocument();
    expect(form).toBeInTheDocument();
    expect(formInput).toBeInTheDocument();
  })

  it('Checks on handleToggle function', () =>{
    // Setup
    const toDoList = data;
    let mapped = [];
    const setToDoList = jest.fn()
    const handleToggle = jest.fn((value) =>{
      mapped = toDoList.map(task => {
        let temp = task.id === Number(value) ? {...task, complete: !task.complete} : {...task}
        return temp
      })
      setToDoList(mapped)
    })

    render(<App />)

    // Assert
    const item = screen.getByText("Do laundry")

    // Action
    fireEvent.click(item)
    handleToggle(item.id)


    // Expectations
    expect(handleToggle).toHaveBeenCalledWith("2");
    expect(setToDoList).toHaveBeenCalledWith(mapped);
    expect(mapped[1].task).toEqual('Do laundry')
    expect(mapped[1].complete).toEqual(false)
  })

  it('Checks the handleCompleted function', () =>{
    let toDoList = data;
    let filtered = [];
    const setToDoList = jest.fn(() => {
      toDoList = filtered;
    });
    const handleCompleted = jest.fn(() => {
      filtered = toDoList.filter(task => {
        return !task.complete
      })
      setToDoList(filtered)
    })
    
    render(<App />)

    // Asserts
    const clearButton = screen.getByText('Clear Completed')

    fireEvent.click(clearButton)
    expect(toDoList).toHaveLength(data.length);
    expect(filtered).toEqual([])
    handleCompleted()
    expect(filtered.map(task => task.complete)).toEqual([
      false,
      false
    ])
    expect(setToDoList).toHaveBeenCalledWith(filtered)
    expect(toDoList).toHaveLength(filtered.length)
  })

  it('Checks the addTask function', () => {
    // Setup
    let toDoList = data;
    let userInput = 'My Task';
    let newTasklist = [];
    const setUserInput = jest.fn();
    const setToDoList = jest.fn();
    const addTask = jest.fn((userInput) =>{
      newTasklist = [...toDoList]
      newTasklist = [...newTasklist, {id: toDoList.length + 1, task: userInput, complete: false}];
  
      // Sort new IDs - prevents items from having the same ID.
      newTasklist = newTasklist.map((task, index) => {
        return {...task, id: index + 1}
      })
      setToDoList(newTasklist)
    })

    const handleSubmit = jest.fn(() => {
      addTask(userInput)
      setUserInput('')
    })

    render(<App />)

    // Asserts
    const form = screen.getByTestId("form")

    // Actions
    fireEvent.submit(form)
    expect(newTasklist).toEqual([])
    expect(toDoList).toHaveLength(data.length);
    handleSubmit()
    expect(newTasklist = [
      ...toDoList,
      {
        id: toDoList.length + 1,
        task: 'My Task',
        complete: false
      }
    ]).toHaveLength(toDoList.length + 1)
    expect(newTasklist.map(task => task.id)).toEqual([
      1,
      2,
      3,
      4,
      5,
      6
    ])
    expect(setToDoList).toHaveBeenCalledWith(newTasklist)
  })
})