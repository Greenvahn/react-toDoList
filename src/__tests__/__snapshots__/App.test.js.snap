// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App.js /> the task data is correct 1`] = `
Array [
  Object {
    "complete": true,
    "id": 1,
    "task": "Give dog a bath",
  },
  Object {
    "complete": true,
    "id": 2,
    "task": "Do laundry",
  },
  Object {
    "complete": false,
    "id": 3,
    "task": "Vacuum floor",
  },
  Object {
    "complete": true,
    "id": 4,
    "task": "Feed cat",
  },
  Object {
    "complete": false,
    "id": 5,
    "task": "Change light bulbs",
  },
]
`;
